chain/chain/src/types.rs

```diff
@@ -575,7 +575,11 @@ pub trait RuntimeAdapter: Send + Sync {
     fn get_protocol_config(&self, epoch_id: &EpochId) -> Result<ProtocolConfig, Error>;
 }
 
-pub trait RuntimeWithEpochManagerAdapter: RuntimeAdapter + EpochManagerAdapter {}
+/// LEGACY trait. Will be removed. Use RuntimeAdapter or EpochManagerHandler instead.
+pub trait RuntimeWithEpochManagerAdapter: RuntimeAdapter + EpochManagerAdapter {
+    fn epoch_manager_adapter(&self) -> &dyn EpochManagerAdapter;
+    fn epoch_manager_adapter_arc(&self) -> Arc<dyn EpochManagerAdapter>;
+}
```

Comment:
As a intermediate step to get from A to B, I can tolerate this. But I think you made your life harder than it needed to be.  :)

In Rust, trait inheritance really isn't inheritance as one expects from OOP and all attempts to use in such way will result in pain. I would have probably chosen a composition strategy, something along the lines of:

```rust
struct RuntimeWithEpochManagerAdapter {
    runtime: Arc<dyn RuntimeAdapter>,
    epoch_manager: Arc<dyn EpochManagerAdapter>,
}
```

Then you can use this struct in places the trait `RuntimeWithEpochManagerAdapter` was used before. 

`trait RuntimeWithEpochManagerAdapter: RuntimeAdapter + EpochManagerAdapter {}` may look like it gives you the same. But it doesn't. It is only a short-hand for the type bound `RuntimeAdapter + EpochManagerAdapter`, which is still checked on each type separately at compile-time and doesn't have a way to cast at runtime.

Reply:
You are totally right. This honestly was my oversight, because I used that exact same pattern you suggested for the TestLoop, but not here - and I used it exactly to solve this problem!

Well nonetheless, thinking about it now, there is a non-trivial amount of boilerplate required with this solution, because any existing user of RuntimeWithEpochManagerAdapter (which should be stripped of the `Arc<dyn >`) would need to either call functions via `.runtime` or `.epoch_manager` (which would be a simple but time-consuming refactoring), or we would need to define forwarding functions on RuntimeWithEpochManagerAdapter, one for each function of RuntimeAdapter and EpochManagerAdapter.

Oh well, I suppose given that we're already here, it perhaps is better to use the time to getting rid of this whole thing faster.